ND100/nd100_cpu.c

```diff
@@ -1,1737 +1,1748 @@
+		ins_lrb(IR);
+	} else if ((IR & ND_MONMSK) == ND_MON) {
+		RBLK[14][rnT] = SEXT8(IR);
+		intrpt14(IIE_MC);
+	} else if ((IR & ND_MONMSK) == ND_WAIT) {
+		if (ISION() == 0) {
+			regP++;
+			reason = STOP_WAIT;
+		} else if (curlvl > 0) {
+			pid &= ~(1 << curlvl);
+			n = highest_level();
+			if (curlvl != n) {
+				regP++;
+				for (i = 0; i < 8; i++) {
+					RBLK[curlvl][i] = R[i];
+					R[i] = RBLK[n][i];
+				}
+				regP--;
+			}
+			curlvl = n;
+		}
+	} else if ((IR & ND_MONMSK) == ND_MIS_NLZ) {
+		/* convert integer to floating point */
+		ins_nlz(IR);
+	} else if ((IR & ND_MONMSK) == ND_MIS_DNZ) {
+		ins_dnz(IR);
+	} else
+		reason = STOP_UNHINS;
+	return reason;
+}
+
+int
+ins_sht(int IR, int off)
+{
+	char sht_reg[] = { rnT, rnD, rnA, 0 };
+	int m, n, rs, i;
+	uint32 ushc;
+
+	rs = sht_reg[(IR >> 7) & 03];
+	n = BIT5(IR) ? 32 - IR & 037 : IR & 037;
+	m = BIT7(regSTL);
+
+	ushc = rs ? R[rs] : (regA << 16) | regD;
+	if (BIT5(IR)) { /* right */
+		int mm = BIT0(ushc);
+		for (i = 0; i < n; i++) {
+			m = BIT0(ushc);
+			ushc = ushc >> 1;
+			switch (IR & 03000) {
+			case 0: /* Arithmetic */
+				ushc |= (rs ? (BIT14(ushc) << 15) :
+				    (BIT30(ushc) << 31));
+				break;
+			case 01000: /* ROT */
+				ushc |= (rs ? (m << 15) :
+				    (m << 31));
+				break;
+			case 02000: /* zero end input */
+				break;
+			case 03000: /* link end input */
+				ushc |= (rs ? (mm << 15) :
+				    (mm << 31));
+				break;
+			}
+		}
+	} else { /* left */
+		int mm = (rs ? BIT15(ushc) : BIT31(ushc));
+		for (i = 0; i < n; i++) {
+			m = (rs ? BIT15(ushc) : BIT31(ushc));
+			ushc = ushc << 1;
+			switch (IR & 03000) {
+			case 01000: /* ROT */
+				ushc |= m; break;
+			case 0: /* Arithmetic */
+			case 02000: /* zero end input */
+				break;
+			case 03000: /* link end input */
+				ushc |= mm; break;
+			}
+		}
+	}
+	regSTL = (regSTL & ~STS_M) | (m << 7);
+	if (rs == 0) {
+		regA = ushc >> 16;
+		regD = ushc;
+	} else
+		 R[rs] = ushc;
+	return SCPE_OK;
+}
+
+int
+ins_na(int IR, int addr)
+{
+	return STOP_UNHINS;
+}
+
+int
+ins_iox(int IR, int addr)
+{
+	return iox_check(IR & ND_IOXMSK);
+}
+
+int
+ins_arg(int IR, int addr)
+{
+	int rs, n = (IR >> 8) & 03;
+
+	rs = n ? n + 4 : 3;
+	R[rs] = add3(BIT10(IR) ? R[rs] : 0, SEXT8(IR), 0);
+	return SCPE_OK;
+}
+
+int
+ins_bop(int IR, int addr) 
+{                       
+
+	int rd, n, reason = 0;
+
+	rd = IR & 7;
+	n = (IR >> 3) & 017;
+	switch ((IR >> 7) & 017) {
+	case 000: /* BSET zero/one */
+		R[rd] &= ~(1 << n);
+		break;
+
+	case 001: 
+		R[rd] |= (1 << n);
+		break;
+
+	case 002: /* BSET BCM bit = ~bit */
+		R[rd] ^= (1 << n);
+		break;
+
+	case 003: /* BSET BAC bit = K */
+		R[rd] &= ~(1 << n);
+		if (regSTL & STS_K)
+			R[rd] |= (1 << n);
+		break;
+
+	case 004: /* BSKP zero/one */
+	case 005: 
+		if (((R[rd] >> n) & 1) == BIT7(IR))
+			regP++;
+		break;
+
+	case 006: /* BSKP BCM K == ~bit */
+		if (((regSTL & STS_K) != 0) ^ (((R[rd] >> n) & 1) != 0))

```

Comment:
Since `^` is a bit operator, not a Boolean operator, it might be clearer to write it as `!=`.  But ok either way.